vmap: added fallback for in-place operators

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46191 vmap: added fallback for in-place operators**
* #46190 Rename createLevelsBitset -> createVmapLevelsBitset; move it

This PR adds a fallback for in-place operators to vmap. We define an
in-place operator to be an operator that operators in-place on its first
argument and returns the first argument.

The "iteration over batch" logic is mostly copied from the out-of-place
vmap fallback. I wanted to try to not copy this but the iteration logic
is pretty entangled with the rest of the logic; one alternative was to
use if/else statements inside batchedTensorForLoopFallback but then
there are ~3-4 different sites where we would need that.

When in-place operations are not possible
=========================================
Sometimes, an in-place operation inside of vmap is not possible. For
example, `vmap(Tensor.add_, (None, 0))(torch.rand(3), torch.rand(B0, 3))`
is not possible because the tensor being written to in-place has size
[3] and the other tensor has size [B0, 3].

We detect if this is the case and error out inside the in-place
fallback.

Test Plan
=========
Added some new tests to `test_vmap.py`.

Differential Revision: [D24335240](https://our.internmc.facebook.com/intern/diff/D24335240)

zou3519 committed:
vmap: added fallback for in-place operators

This PR adds a fallback for in-place operators to vmap. We define an
in-place operator to be an operator that operators in-place on its first
argument and returns the first argument.

The "iteration over batch" logic is mostly copied from the out-of-place
vmap fallback. I wanted to try to not copy this but the iteration logic
is pretty entangled with the rest of the logic; one alternative was to
use if/else statements inside batchedTensorForLoopFallback but then
there are ~3-4 different sites where we would need that.

When in-place operations are not possible
=========================================
Sometimes, an in-place operation inside of vmap is not possible. For
example, `vmap(Tensor.add_, (None, 0))(torch.rand(3), torch.rand(B0, 3))`
is not possible because the tensor being written to in-place has size
[3] and the other tensor has size [B0, 3].

We detect if this is the case and error out inside the in-place
fallback.

Test Plan
=========
Added some new tests to `test_vmap.py`.

[ghstack-poisoned]

ezyang commented on aten/src/ATen/BatchedFallback.cpp:
```diff
@@ -34,6 +34,172 @@ static bool areAnyArgumentsTensorList(const FunctionSchema& schema) {
+// 4. No other arguments are aliased
+// Here is an example of an in-place operator:
+// add_(Tensor(a!) self, Tensor other, *, Scalar alpha=1) -> Tensor(a!)
+static bool isInplaceOp(const c10::FunctionSchema& schema) {
```
You don't have to fix this here, but this should be precomputed in FunctionSchema, as the inplace-ness is of general interest to others. In fact, I have a version of this function in conjugate view tensors: https://github.com/pytorch/pytorch/pull/44799/files#diff-e0b7fe2a6c006c34bd33d0808fe1c9f4659d0faf27f58d64758c96be46c9ac77R31 though I am not only testing for inplaceness, I am testing for any mutation at all.

One Q I will be trying to answer as I read the rest of this PR is whether or not you really needed this algorithm to only work on inplace operations, or if any mutating operation would have been OK.

zou3519 replied:
The algorithm assumes that the first argument of the function is the only one that is being mutated. It is possible to extend the algorithm to support functions that mutate a different argument (like out=) or mutate multiple arguments, but those require some more bookkeeping (and the plan for vmap is to not support out= overloads)

zou3519 replied:
I'm happy to toss something like this into FunctionSchema, in this PR or in a separate PR

ezyang replied:
out= doesn't seem like a hard generalization. Besides for safety checks, the mutating argument(s) simply just may occur in other positions.

Of course, boxed fallback for out functions doesn't work yet (it's buggy), so you couldn't implement it anyhow yet.

ezyang commented on aten/src/ATen/BatchedFallback.cpp:
```diff
@@ -34,6 +34,172 @@ static bool areAnyArgumentsTensorList(const FunctionSchema& schema) {
+// Here is an example of an in-place operator:
+// add_(Tensor(a!) self, Tensor other, *, Scalar alpha=1) -> Tensor(a!)
+static bool isInplaceOp(const c10::FunctionSchema& schema) {
+  if (!schema.is_mutable() || schema.name().back() != '_') {
```
The underscore test here is not sound; double underscore operations indicate inplaceness by being `__iblah__`. The current canonical definition of inplaceness is in BaseOperatorName in tools/codegen/model.py

zou3519 replied:
Thanks for the catch! There's also schemas like `__ior__` that do in-place (as declared by the `i`)... I'll take a look at tools/codegen/model.py and try to copy that

zou3519 replied:
Deleted the schema.name().back() != '_' check.

ezyang commented on aten/src/ATen/BatchedFallback.cpp:
```diff
@@ -34,6 +34,172 @@ static bool areAnyArgumentsTensorList(const FunctionSchema& schema) {
+// The general flow of the algorithm is as follows.
+// - First, we figure out which arguments are BatchedTensors and save them
+//   to a vector. We also store a vector of which index of the arguments list
+//   each BatchedTensor appears in. This will be useful for bookkeeping later.
```
Wondering if you actually need to extract out the BatchedTensors here or not; maybe a uniform representation which abstracts over batched tensors and non-batched tensors would make the algorithm here more uniform.

zou3519 replied:
A uniform representation would save us from storing a vector of indices each BatchedTensor appears in. We could do the following:
- save all Tensor arguments to a vector
- vmap transform all of the BatchedTensors in the Tensor arguments
    - Right now, vmap transforms assume that their inputs are BatchedTensors. We'd have to make the vmap transform API more uniform.
- While iterating through all of the arguments, insert transformed tensors instead of the original tensors onto the stack to be operated on.

I guess we already have a uniform representation - a Tensor can either be batched or not-batched. To make the uniform representation work out we'd have to go and refactor vmap transforms to work with them

ezyang replied:
Because a non-batched tensor is "simply" a batch tensor with no vmap dims.

zou3519 replied:
Yup, exactly. I think a refactor of the vmap transforms API is in store; I've been trying to figure out what it should look like

ezyang commented on aten/src/ATen/BatchedFallback.cpp:
```diff
@@ -34,6 +34,172 @@ static bool areAnyArgumentsTensorList(const FunctionSchema& schema) {
+      TORCH_CHECK(false,
+        "vmap: ", schema.name(), "(tensor, ...) is not possible because `tensor` ",
+        "is not being vmapped over for vmap level ", offending_level, " but another "
+        "argument is. Please try to use out-of-place operators instead of ",
```
Your comment above (`other` has more elements elements than `self` due to being vmapped over) was pretty clear, but I'm not sure if I would have gotten the same gist from this error message.

zou3519 replied:
That's a good point, I'll try to rephrase this to be in terms of elements rather than vmap levels

ezyang commented on aten/src/ATen/BatchedFallback.cpp:
```diff
@@ -34,6 +34,172 @@ static bool areAnyArgumentsTensorList(const FunctionSchema& schema) {
+    batched_tensor_inputs.push_back(tensor);
+    batched_tensor_inputs_position.push_back(idx);
+  }
+  TORCH_INTERNAL_ASSERT(batched_tensor_inputs.size() > 0);
```
Can't I violate this invariant with an inplace operation like `x.neg_()`?

zou3519 replied:
`x.neg_()` pushes a single argument onto the stack, x, so batched_tensor_inputs will have size 1 in this case

ezyang commented on aten/src/ATen/BatchedFallback.cpp:
```diff
@@ -34,6 +34,172 @@ static bool areAnyArgumentsTensorList(const FunctionSchema& schema) {
+
+  // Compute the total number of batches
+  auto num_batch_dims = input_physical_views.front().numBatchDims();
+  auto some_sizes = input_physical_views.front().tensor().sizes();
```
"some" sizes? :O

zou3519 replied:
My naming creativity ran out of ideas lol. I'll spend more time thinking of a better name

ezyang commented on aten/src/ATen/BatchedFallback.cpp:
```diff
@@ -34,6 +34,172 @@ static bool areAnyArgumentsTensorList(const FunctionSchema& schema) {
+
+static int64_t firstSetBitPosition(std::bitset<kVmapMaxTensorDims> bdims) {
+  auto bdims_as_int64 = bdims.to_ulong();
+  return std::log2(bdims_as_int64 & -bdims_as_int64) + 1;
```
We have a direct implementation of this in `c10/util/llvmMathExtras.h`, specifically findLastSet

ezyang commented on aten/src/ATen/BatchedFallback.cpp:
```diff
@@ -50,17 +216,21 @@ static bool areAnyArgumentsTensorList(const FunctionSchema& schema) {
   const auto& schema = op.schema();
   const auto num_returns = schema.returns().size();
+
+  if (isInplaceOp(schema)) {
```
https://github.com/pytorch/pytorch/issues/45203 may be of interest here

zou3519 replied:
Yes, that is definitely useful to avoid recomputation here

zou3519 committed:
Update on "vmap: added fallback for in-place operators"


This PR adds a fallback for in-place operators to vmap. We define an
in-place operator to be an operator that operators in-place on its first
argument and returns the first argument.

The "iteration over batch" logic is mostly copied from the out-of-place
vmap fallback. I wanted to try to not copy this but the iteration logic
is pretty entangled with the rest of the logic; one alternative was to
use if/else statements inside batchedTensorForLoopFallback but then
there are ~3-4 different sites where we would need that.

When in-place operations are not possible
=========================================
Sometimes, an in-place operation inside of vmap is not possible. For
example, `vmap(Tensor.add_, (None, 0))(torch.rand(3), torch.rand(B0, 3))`
is not possible because the tensor being written to in-place has size
[3] and the other tensor has size [B0, 3].

We detect if this is the case and error out inside the in-place
fallback.

Test Plan
=========
Added some new tests to `test_vmap.py`.

[ghstack-poisoned]
